### PR TITLE
feat: #638 Add cronjob config for database restore

### PR DESCRIPTION
**Pull request by MCatherine1994**
- added a yaml file for setting up a cronjob for database restore, refs: #638

Full documentation is at: https://github.com/bcgov/nr-fom/wiki/Database-Restore

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-24.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-24.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-24.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
 - [api](https://fom-24.apps.silver.devops.gov.bc.ca/api)
 - [admin](https://fom-24.apps.silver.devops.gov.bc.ca/admin)
 - [public](https://fom-24.apps.silver.devops.gov.bc.ca/public)

Once merged, code will be promoted and handed off to following workflow run.
 - [Main Merge Workflow](https://github.com/bcgov/nr-fom/actions/workflows/merge-main.yml)

### CODE REVIEW

File: db/openshift.deploy.yml [hunk 1]
```diff
@@ -326,3 +329,86 @@ objects:
...
+          template:
+            metadata:
+              labels:
+                app: ${NAME}-${ZONE}
+                cronjob: ${NAME}-${ZONE}
+            spec:
+              containers:
+                - name: ${NAME}-${ZONE}-${COMPONENT}-restore
+                  # use the same image as our database, so we can run the psql command
+                  image: image-registry.apps.silver.devops.gov.bc.ca/${OC_NAMESPACE}/${NAME}-${ZONE}-${COMPONENT}:${ZONE}-db
```
Comment (MCatherine1994) [inline db/openshift.deploy.yml:360]: add the new OC_NAMESPACE variable because we're having problem to pull images before, I got the same error. So here we pull the image from this url

File: db/openshift.deploy.yml [hunk 2]
```diff
@@ -326,3 +329,86 @@ objects:
...
+                  command: ["/bin/sh", "-c"]
+                  args:
+                  - |
+                    sql_zip_file=$(find backups/${RESTORE_DIR} -type f -name "*.sql.gz" -print0)
```
Comment (MCatherine1994) [inline db/openshift.deploy.yml]: backup files created by Openshift backup container are zipped sql files, so we need to unzip it first

File: .github/workflows/merge-main.yml [hunk 3]
```diff
@@ -32,7 +32,7 @@ jobs:
             oc_version: "4.13"
             overwrite: true
             parameters:
-              -p OC_NAMESPACE=a4b31c-test
+              -p OC_NAMESPACE=${{ vars.OC_NAMESPACE }}
```
Comment (basilv) [inline .github/workflows/merge-main.yml:35]: I think this needs to be test and the one below needs to be prod (different values), but I didn't look at this closely so could be wrong.

File: db/openshift.deploy.yml (hunk 1 above)
Comment (basilv) [reply db/openshift.deploy.yml:360]: Why are you defining a cron job to test database restore? We don't want this running normally, especially not in production. But perhaps I'm misunderstanding the scope of what this restore process is trying to do. What I'd like to see first before these changes is the documented database restore process (in Confluence). That's what we're trying to test.

File: db/openshift.deploy.yml [hunk 4]
```diff
@@ -326,3 +329,86 @@ objects:
...
+                    echo "Found zipped backup file: $sql_zip_file"
+                    gzip -d $sql_zip_file
+                    sql_file=$(find backups/${RESTORE_DIR} -type f -name "*.sql" -print0)
```
Comment (basilv) [inline db/openshift.deploy.yml]: Which backup file is being selected here? Not clear to me.

File: .github/workflows/pr-open.yml [hunk 5]
```diff
@@ -106,7 +106,8 @@ jobs:
             triggers: ('db/' 'libs/' 'api/' 'admin/')
           - name: db
             file: db/openshift.deploy.yml
-            overwrite: false
+            overwrite: true
```
Comment (basilv) [inline .github/workflows/pr-open.yml]: Why is this change necessary?

File: .github/workflows/merge-main.yml [hunk 6]
```diff
@@ -97,7 +98,8 @@ jobs:
             parameters: -p URL=fom.nrs.gov.bc.ca
           - name: db
             file: db/openshift.deploy.yml
-            overwrite: false
+            overwrite: true
```
Comment (basilv) [inline .github/workflows/merge-main.yml]: I'm nervous about changing this overwrite configuration for prod.

File: db/openshift.deploy.yml [hunk 7]
```diff
@@ -326,3 +329,86 @@ objects:
...
+                        secretKeyRef:
+                          name: ${NAME}-${ZONE}-${COMPONENT}
+                          key: database-user
+                    # we don't need to declare the restore_dir variable if we set it manually, just put here so we won't forget
+                    # restore_dir need to be manually set, in the format for example, daily/2024-08-16, where backup container puts the backup files
+                    - name: RESTORE_DIR
+                      value: ""
+                    # we don't need to declare the target_host variable if we set it manually, just put here so we won't forget
+                    # target_host is the ip address of the new database pod
+                    - name: TARGET_HOST
```
Comment (basilv) [inline db/openshift.deploy.yml]: I don't get this comment - don't we have to set TARGET_HOST? Not sure why we're spinning up a new database pod in a restore scenario - how is the application being repointed then at the restored database?

File: .github/workflows/merge-demo.yml [hunk 8]
```diff
@@ -55,7 +55,7 @@ jobs:
             file: api/openshift.deploy.yml
             overwrite: true
             parameters:
-              -p OC_NAMESPACE=a4b31c-test
+              -p OC_NAMESPACE=${{ vars.OC_NAMESPACE }}
```
Comment (ianliuwk1019) [inline .github/workflows/merge-demo.yml:58]: Hi Catherine, just curious because I can't verify this. 
Is this environment variable from github? Value exists at github for test and prod I assume; and is based on 'environment' in template for different values? Why is this change necessary?

File: .github/workflows/pr-open.yml [hunk 9]
```diff
@@ -106,7 +106,8 @@ jobs:
...
+            parameters: -p OC_NAMESPACE=${{ vars.OC_NAMESPACE }}
```
Comment (ianliuwk1019) [inline .github/workflows/pr-open.yml:110]: Just for my curiosity.. I don't see 'environment' in template, so this `vars.OC_NAMESPACE` will get the value "a4b31c-test"?

File: db/openshift.deploy.yml [hunk 10]
```diff
@@ -326,3 +329,86 @@ objects:
               dnsPolicy: "ClusterFirst"
               serviceAccountName: "${JOB_SERVICE_ACCOUNT}"
               serviceAccount: "${JOB_SERVICE_ACCOUNT}"
+  - kind: CronJob
+    apiVersion: "batch/v1"
```
Comment (ianliuwk1019) [inline db/openshift.deploy.yml:333]: Maybe just "v1"?

File: .github/workflows/merge-demo.yml (hunk 8 above)
Comment (MCatherine1994) [reply .github/workflows/merge-demo.yml:58]: yes, this is environment variable, I just feel we don't need to hardcode it. But it's unrelated to database restore, I just noticed it. I can change it back if we feel unsafe. But ideally we should use the variable, not hardcode it.

Screen Shot 2024-08-20 at 9 37 59 AM

File: .github/workflows/merge-main.yml (hunk 3 above)
Comment (MCatherine1994) [reply .github/workflows/merge-main.yml:35]: This is Github environment variable. I just feel we don't need to hardcode it. But it's unrelated to database restore, I just noticed it. I can change it back if we feel unsafe. But ideally we should use the variable, not hardcode it.
image

File: .github/workflows/merge-main.yml (hunk 6 above)
Comment (MCatherine1994) [reply .github/workflows/merge-main.yml]: I think this is just let Derek's GitHub Action know we will pass some parameters, but I could try to set it to false.

File: db/openshift.deploy.yml (hunk 4 above)
Comment (ianliuwk1019) [reply db/openshift.deploy.yml]: If I understand correctly, is the `${RESTORE_DIR}` same as directory used for backup file for daily/weekly? And the same directory is used for unzip the backup file?

File: .github/workflows/pr-open.yml (hunk 9 above)
Comment (MCatherine1994) [reply .github/workflows/pr-open.yml:110]: this will be dev, if we didn't specify the environment, then it will use the default Github repository environment variable, and the value of `OC_NAMESPACE` is "a4b31c-dev" as repository variable. If we specify environment, it will use GitHub environment variable.

File: db/openshift.deploy.yml (hunk 7 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: So we will need to manually setup these values in order to trigger this job. The target host needs to be the IP address of the database pod, which could change whenever the database redeploy.

That's also why I feel it's okay to have a cronjob. Even someone triggers the job, without setting up these values, the job will fail. 

Please ignore my spinning up new database scenario for now. I have another idea and need to explain that.

File: db/openshift.deploy.yml [hunk 11]
```diff
@@ -326,3 +329,86 @@ objects:
...
+                    sql_file=$(find backups/${RESTORE_DIR} -type f -name "*.sql" -print0)
+                    echo "Running SQL file: $sql_file"
+                    psql -h ${TARGET_HOST} -U ${POSTGRES_USER} -d ${POSTGRES_DB} -f $sql_file
+                    gzip $sql_file
+                    echo "Finish database restore"
+                  volumeMounts:
+                    - name: ${NAME}-${ZONE}-${COMPONENT}-restore
+                      mountPath: "${BACKUP_DIR}"
+                    - name: ${NAME}-${ZONE}-${COMPONENT}-restore
+                      mountPath: /var/lib/postgresql
```
Comment (ianliuwk1019) [inline db/openshift.deploy.yml]: Maybe only because I am not familiar so get confused... why are two 'mountPath's on the same volume "${NAME}-${ZONE}-${COMPONENT}-restore"?

File: db/openshift.deploy.yml (hunk 4 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: restore directory we will set manually, depends on which backup we want to restore

File: db/openshift.deploy.yml (hunk 10 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml:333]: Hi Ian, what's the difference between batch/v1 and v1?

Comment (basilv) [reply db/openshift.deploy.yml:333]: OpenShift expects a specific version value, we can't just change it arbitrarily. "batch/v1" is correct.

File: db/openshift.deploy.yml (hunk 11 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: I got the same confusion before, but I can only give the same name, whatever other names I tried, it will return me an error. So I think it's because the volume for this container is named `${NAME}-${ZONE}-${COMPONENT}-restore`, and we need to mount both backup directory to get all the backup files, and mount the Postgres directory to get the Postgres command line tool
Screen Shot 2024-08-20 at 10 16 40 AM

File: db/openshift.deploy.yml [hunk 12]
```diff
@@ -326,3 +329,95 @@ objects:
...
+                    sql_file=$(find backups/${BACKUP_FILE_DIR} -type f -name "*.sql" -print0)
+                    if [  -z $sql_file ]; then
+                      echo "No backup SQL file is found"
+                    else
+                      echo "Found SQL file, rename existing database, and create a new empty database"
+                      psql -h ${NAME}-${ZONE}-${COMPONENT} -U ${POSTGRES_USER} -c "ALTER DATABASE fom RENAME TO ${RENAME_DATABASE_TO};" -c "CREATE DATABASE fom;"
+                      echo "Running SQL file: $sql_file"
+                      psql -h ${NAME}-${ZONE}-${COMPONENT} -U ${POSTGRES_USER} -d ${POSTGRES_DB} -f $sql_file
+                      echo "Finish database restore"
+                    fi
```
Comment (MCatherine1994) [inline db/openshift.deploy.yml:412]: So now the logic is that:

1. Try to find the sql file first (in case already unzip it in the past)
2. If not found, try to find the zipped sql file, and unzip it
3. Get the sql file again, if not found then print "no backup sql found" and do nothing
4. If found then run restore script

File: db/openshift.deploy.yml [hunk 13]
```diff
@@ -326,3 +329,95 @@ objects:
...
+                        echo "Found zipped backup file: $sql_zip_file"
+                        gzip -dk $sql_zip_file
+                      fi
+                    fi
+                    sql_file=$(find backups/${BACKUP_FILE_DIR} -type f -name "*.sql" -print0)
+                    if [  -z $sql_file ]; then
+                      echo "No backup SQL file is found"
+                    else
+                      echo "Found SQL file, rename existing database, and create a new empty database"
+                      psql -h ${NAME}-${ZONE}-${COMPONENT} -U ${POSTGRES_USER} -c "ALTER DATABASE fom RENAME TO ${RENAME_DATABASE_TO};" -c "CREATE DATABASE fom;"
```
Comment (MCatherine1994) [inline db/openshift.deploy.yml]: We can use the database service name as the host, also I think we can manually setup the name for how we want to rename the existing "fom" database.

File: db/openshift.deploy.yml (hunk 11 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: removed the PostgreSQL mount as no need

File: db/openshift.deploy.yml [hunk 14]
```diff
@@ -326,3 +329,95 @@ objects:
...
+                          key: database-password
+                    - name: POSTGRES_USER
+                      valueFrom:
+                        secretKeyRef:
+                          name: ${NAME}-${ZONE}-${COMPONENT}
+                          key: database-user
+                    # we don't need to declare the BACKUP_FILE_DIR variable if we set it manually, just put here so we won't forget
+                    # BACKUP_FILE_DIR need to be set manually, in the format for example, daily/2024-08-16, where backup container puts the backup files
+                    - name: BACKUP_FILE_DIR
+                      value: ""
```
Comment (MCatherine1994) [inline db/openshift.deploy.yml:439]: also rename this, so call it "BACKUP_FILE_DIR" maybe this sounds better?

File: db/openshift.deploy.yml [hunk 15]
```diff
@@ -326,3 +329,95 @@ objects:
...
+                    # RENAME_DATABASE_TO need to be set manually, rename the existing fom database to a new name
+                    - name: RENAME_DATABASE_TO
```
Comment (MCatherine1994) [inline db/openshift.deploy.yml]: Hi Ian, how we want to name this variable? basically this will be the name that we use to rename the existing "fom" database to.... I don't know what it should be called...

Comment (ianliuwk1019) [reply db/openshift.deploy.yml]: Hi Catherine, I have no preference for the name. Does fom_old sound good?

File: .github/workflows/pr-open.yml (hunk 5 above)
Comment (MCatherine1994) [reply .github/workflows/pr-open.yml]: seems no need, switched back

File: db/openshift.deploy.yml (hunk 15 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: how about OLD_FOM_DATABASE_NAME?

Comment (basilv) [reply db/openshift.deploy.yml]: I think this should have a default value...

Comment (basilv) [reply db/openshift.deploy.yml]: (so it doesn't need to be specified every time).

File: db/openshift.deploy.yml [hunk 16]
```diff
@@ -326,3 +329,95 @@ objects:
...
+      concurrencyPolicy: "Replace"
+      successfulJobsHistoryLimit: ${{SUCCESS_JOBS_HISTORY_LIMIT}}
+      failedJobsHistoryLimit: ${{FAILED_JOBS_HISTORY_LIMIT}}
+      jobTemplate:
+        metadata:
+          labels:
+            app: ${NAME}-${ZONE}
+            cronjob: ${NAME}-${ZONE}
+        spec:
+          backoffLimit: ${{JOB_BACKOFF_LIMIT}}
```
Comment (basilv) [inline db/openshift.deploy.yml]: Where are these limits set?

File: db/openshift.deploy.yml (hunk 12 above)
Comment (basilv) [reply db/openshift.deploy.yml:412]: Should exit with error if no backup SQL file is found. Should check the return value of the psql commands and report errors.

Comment (basilv) [reply db/openshift.deploy.yml:412]: The restore finishes with the old schema existing, which is fine. Part of the recovery process could be to remove that. Also, maybe this process should perform a backup first before doing the restore?

**Review by basilv — CHANGES_REQUESTED** [review]

File: db/openshift.deploy.yml (hunk 16 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: these are set for the db backup cronjob, I thought maybe I can use them here as well

File: db/openshift.deploy.yml (hunk 15 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: yeah I was using a default value before. And while doing testing, if I want to run the restore multiple times, I need to manually remove the renamed database, otherwise I'll got the error while renaming the database again, cause it already exists from the previous restore. So I was thinking maybe we can set the name manually, but I'm fine with a default value.

File: db/openshift.deploy.yml (hunk 12 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml:412]: yeah, we should remove that, otherwise we can't run the restore multiple times. Yeah, I like the idea to the cleanup before run the restore, thanks!!

Comment (MCatherine1994) [reply db/openshift.deploy.yml:412]: when no backup file is found, it will just print "Not found" and do nothing for now. I'll check how to report errors.
Screen Shot 2024-08-20 at 2 41 06 PM

File: db/openshift.deploy.yml (hunk 16 above)
Comment (basilv) [reply db/openshift.deploy.yml]: Hmm, not sure if that makes sense, since we have retry = never.

File: db/openshift.deploy.yml (hunk 12 above)
Comment (basilv) [reply db/openshift.deploy.yml:412]: So rather than having the cron job finish successfully in the case of no backup SQL file found, we should have the job run marked as a failure (which I assume will happen with a non-zero exit code from the script).

File: db/openshift.deploy.yml (hunk 16 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: um... this I'm not so sure... just follow the one for db backup

File: db/openshift.deploy.yml (hunk 12 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml:412]: When no backup file found, exit 1:
Screen Shot 2024-08-21 at 4 05 42 PM

When psql command failed, exit 1 (the following example failure happened when I pass an empty file `psql -h fom-24-db -U ${POSTGRES_USER} -d ${POSTGRES_DB} -f ''`):

Screen Shot 2024-08-21 at 4 12 54 PM

Comment (MCatherine1994) [reply db/openshift.deploy.yml:412]: I think it's better to run backup before restore, but not sure if we want to include that in this script as well, it's getting complicated....

Comment (basilv) [reply db/openshift.deploy.yml:412]: Yes, probably sufficient for now to just have our documented process say to do a manual backup before trying to do the restore. (and bring the app down before doing the backup).

File: db/openshift.deploy.yml (hunk 16 above)
Comment (MCatherine1994) [reply db/openshift.deploy.yml]: I changed to 0. The restart policy sets to never will ensure when the pod fails, it won't be restarted. But we still need to set backofflimit to make sure the job will not be retired.

**Comment by basilv** [comment]
Reading the wiki process: "Do a manual backup of the current database (can do that in the database pod, so this temporary backup will be stored in the database volume)" -> the specific commands to run should be provided. I'm a little concerned by doing this versus the normal backup process as if you need to restore from this special backup you need a special restore process, but on the other hand if the regular restore process fails, this approach should still succeed. Just want to be careful that if this special backup isn't stored within a PVC, then if the DB pod restarts the backup will be lost.

File: db/openshift.deploy.yml [hunk 17]
```diff
@@ -326,3 +329,99 @@ objects:
...
+              labels:
+                app: ${NAME}-${ZONE}
+                cronjob: ${NAME}-${ZONE}
+            spec:
+              containers:
+                - name: ${NAME}-${ZONE}-${COMPONENT}-restore
+                  # use the same image as our database, so we can run the psql command
+                  image: image-registry.apps.silver.devops.gov.bc.ca/${OC_NAMESPACE}/${NAME}-${ZONE}-${COMPONENT}:${ZONE}-db
+                  command: ["/bin/sh", "-c"]
+                  args:
```
Comment (basilv) [inline db/openshift.deploy.yml:362]: I thought we talked about making this a script file that's loaded into the database image. But I'm okay if it is left like this.

File: db/openshift.deploy.yml [hunk 18]
```diff
@@ -326,3 +329,99 @@ objects:
...
+                        gzip -dk $sql_zip_file
+                      fi
+                    fi
+                    sql_file=$(find backups/${BACKUP_FILE_DIR} -type f -name "*.sql" -print0)
+                    if [  -z $sql_file ]; then
+                      echo "Error: No backup SQL file is found"
+                      exit 1
+                    else
+                      echo "Found SQL file, rename existing database, and create a new empty database"
+                      psql -h ${NAME}-${ZONE}-${COMPONENT} -U ${POSTGRES_USER} -c "DROP DATABASE IF EXISTS ${OLD_FOM_DATABASE_NAME};" -c "ALTER DATABASE fom RENAME TO ${OLD_FOM_DATABASE_NAME};" -c "CREATE DATABASE fom;"
```
Comment (basilv) [inline db/openshift.deploy.yml]: Still not checking return code from this psql call.

Comment (basilv) [inline db/openshift.deploy.yml]: The process documented in the wiki could talk more about what to do if the restore fails (or wasn't the correct restore point and needs to be redone), which could mean picking a different value for old_fom_database_name, and/or one option for reverting the restore would be to drop the fom database and rename old_fom to fom...

**Comment by webgismd** [comment]
Team Evergreen has in their backlog to look at DB restore/backup as part of STRA-- goal was to not use PVC but object storage. they had buckets created for this purpose, but I am not sure how far they have gone with it yet. @craigyu or @DerekRoberts may know?

**Comment by DerekRoberts** [comment]
@webgismd Yup! @RMCampos is working on that. I'll be sidekicking and making it easier to repeat.

**Comment by basilv** [comment]
@webgismd for FOM at least, the database is small and we store a limited number of backups so storage demands are limited. But there are some nice aspects to using object storage instead of a PVC - better resistance to ransomware-style attacks, especially if you can structure your object storage permissions to be insert-only... I'd be more worried though about every team developing a completely different backup/restore process, versus having a defined method for OpenShift Postgres DBs that all the teams can leverage.

**Comment by webgismd** [comment]
I 100% agree here @basilv , FDS is essentially on its own to develop a pattern for itself.  I don't see alot of tactile support from other teams in this area. Perhaps something to discuss with @RMCampos @jazzgrewal @craigyu @paulushcgcj @abschwenker @DerekRoberts

**Comment by DerekRoberts** [comment]
@webgismd @basilv Big yes!  Let's solve this and get it out there.  Our teams, quickstart, developer.gov, stink overflow, etc.

**Comment by paulushcgcj** [comment]
> @webgismd for FOM at least, the database is small and we store a limited number of backups so storage demands are limited. But there are some nice aspects to using object storage instead of a PVC - better resistance to ransomware-style attacks, especially if you can structure your object storage permissions to be insert-only... I'd be more worried though about every team developing a completely different backup/restore process, versus having a defined method for OpenShift Postgres DBs that all the teams can leverage.

The base here was derived from client. In our case, we will keep both a PVC and an S3 backup, as the PVC is a "hot" copy, that will be the more recent one while the S3 will have all past and all current ones, but will be handled as a "cold" copy. This was at least the idea/suggestion I gave and is what we're aiming to do on client.

File: db/openshift.deploy.yml (hunk 18 above)
Comment (ianliuwk1019) [reply db/openshift.deploy.yml]: Adjusted script for return code check and discussed with Basil.

Comment (ianliuwk1019) [reply db/openshift.deploy.yml]: Basic reverting restore failure is included in adjusted script.

**Review by basilv — APPROVED** [review]
